service/featuregate/gates_test.go

```diff
@@ -0,0 +1,72 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package featuregate
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRegistry(t *testing.T) {
+	r := registry{gates: map[string]Gate{}}
+
+	gate := Gate{
+		ID:          "foo",
+		Description: "Test Gate",
+		Enabled:     true,
+	}
+
+	assert.Empty(t, r.list())
+	assert.False(t, r.isEnabled(gate.ID))
+
+	assert.NoError(t, r.add(gate))
+	assert.Len(t, r.list(), 1)
+	assert.True(t, r.isEnabled(gate.ID))
+
+	r.apply(map[string]bool{gate.ID: false})
+	assert.False(t, r.isEnabled(gate.ID))
+
+	assert.Error(t, r.add(gate))
+}
+
+func TestGlobalRegistry(t *testing.T) {
+	resetForTest()
```

Comment:
I think this can cause problems if another package is running a test that relies on a feature that is default-enabled (since tests can be run in parallel). This will cause that other package to see that the gate is disabled (unexpectedly) and fail.

Instead of blanket resetting here I think we can test for individual test-only gates enabled and disabling (disabling available only via non-exported func), but should avoid full reset of the gates in the test or try to verify the full list of the gates.